From 4bd11632aff22f7401db93c9309858cc46b0aa76 Mon Sep 17 00:00:00 2001
From: Timple Raj M <quic_timple@quicinc.com>
Date: Mon, 8 Apr 2024 09:30:57 +0530
Subject: [PATCH] drivers: usb: dwc3: core: Direct update GUCTL and FLADJ
 registers

This patch directly writes the value configured in dts to
the GUCTL and FLADJ registers.

This configuration needed only when the GUCTL and FLADJ values
obtained from dwc3_ref_clk_period function does not meet the
expected values.

dwc3 {
	snps,quirk-fladj = <value>;
	snps,quirk-guctl = <value>;
};

Change-Id: Id7498b898160815c531690e0d311a8be511b2785
Signed-off-by: Timple Raj M <quic_timple@quicinc.com>
---
 drivers/usb/dwc3/core.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/drivers/usb/dwc3/core.c b/drivers/usb/dwc3/core.c
index d25490965b27..afef62e32ba3 100644
--- a/drivers/usb/dwc3/core.c
+++ b/drivers/usb/dwc3/core.c
@@ -359,6 +359,17 @@ int dwc3_core_soft_reset(struct dwc3 *dwc)
 	return 0;
 }
 
+static void dwc3_configure_fladj_guctl(struct dwc3 *dwc)
+{
+	struct device		*dev = dwc->dev;
+	u32 fladj, guctl;
+
+	if (!device_property_read_u32(dev, "snps,quirk-guctl", &guctl))
+		dwc3_writel(dwc->regs, DWC3_GUCTL, guctl);
+	if (!device_property_read_u32(dev, "snps,quirk-fladj", &fladj))
+		dwc3_writel(dwc->regs, DWC3_GFLADJ, fladj);
+}
+
 /*
  * dwc3_frame_length_adjustment - Adjusts frame length if required
  * @dwc3: Pointer to our controller context structure
@@ -1277,6 +1288,8 @@ static int dwc3_core_init(struct dwc3 *dwc)
 	/* Adjust Reference Clock Period */
 	dwc3_ref_clk_period(dwc);
 
+	dwc3_configure_fladj_guctl(dwc);
+
 	dwc3_set_incr_burst_type(dwc);
 
 	ret = dwc3_phy_power_on(dwc);
-- 
2.34.1

